### PR TITLE
[SPARK-14865][SQL] Better error handling for view creation.

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?

This patch improves error handling in view creation. CreateViewCommand itself will analyze the view SQL query first, and if it cannot successfully analyze it, throw an AnalysisException.

In addition, I also added the following two conservative guards for easier identification of Spark bugs:
1. If there is a bug and the generated view SQL cannot be analyzed, throw an exception at runtime. Note that this is not an AnalysisException because it is not caused by the user and more likely indicate a bug in Spark.
2. SQLBuilder when it gets an unresolved plan, it will also show the plan in the error message.

I also took the chance to simplify the internal implementation of CreateViewCommand, and _removed_ a fallback path that would've masked an exception from before.
## How was this patch tested?
1. Added a unit test for the user facing error handling.
2. Manually introduced some bugs in Spark to test the internal defensive error handling.
3. Also added a test case to test nested views (not super relevant).

### CODE REVIEW

**Comment by rxin** [comment]
cc @gatorsmile

**Comment by rxin** [comment]
also cc @liancheng who wrote some of the original code.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 1]
```diff
@@ -85,68 +88,74 @@ case class CreateViewCommand(
...
-        prepareTable(sqlContext, analzyedPlan), ignoreIfExists = false)
+        prepareTable(sqlContext, analyzedPlan), ignoreIfExists = false)
     }
 
     Seq.empty[Row]
   }
 
-  private def prepareTable(sqlContext: SQLContext, analzyedPlan: LogicalPlan): CatalogTable = {
-    val expandedText = if (sqlContext.conf.canonicalView) {
-      try rebuildViewQueryString(sqlContext, analzyedPlan) catch {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: cc @liancheng @yhuai 

I don't think we should ever catch an exception and silently fallback to some other codepath. Let's make sure we don't do that in the future ...

**Comment by yhuai** [comment]
LGTM. Merging to master. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 2]
```diff
@@ -85,68 +88,74 @@ case class CreateViewCommand(
...
-    val logicalPlan = if (tableDesc.schema.isEmpty) {
-      analzyedPlan
-    } else {
-      val projectList = analzyedPlan.output.zip(tableDesc.schema).map {
-        case (attr, col) => Alias(attr, col.name)()
-      }
-      sqlContext.executePlan(Project(projectList, analzyedPlan)).analyzed
-    }
-    new SQLBuilder(logicalPlan).toSQL
+    tableDesc.copy(schema = viewSchema, viewText = Some(viewSQL))
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Everything looks great to me. I only have a comment that is not directly related to this PR.

It looks a little bit strange to generate a partial `CatalogTable` for a `view` in [`createView`](https://github.com/apache/spark/blob/c06110187b3e41405fc13aba367abdd4183ed9a6/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala#L1164). When we describe this parameter `tableDesc` in `CreateViewCommand`, we said it is `the catalog table`. Actually, this is not right. Then, here, we generate the final `CatalogTable` by updating two of the values `schema` and `viewText` with the correct values. 

Anyway, this is not a big deal at all. : )

**Pull request closed without merging**